### ISSUE TITLE
Mainsail and Moonraker for Flashforge AD5M

### ISSUE THREAD

**Issue by xblax** [g992/flashforge-ad5m-5mpro-research#6]
Let's talk a bit about the options to get Mainsail and Moonraker working with the Flashforge 5M (Pro). I think we could consider the following options:

a) Install Mainsail+Moonraker on the Flashforge 5M after rooted into the existing image. As I understand the 128MB ram of the integrated ARM board are a bit limited to run Mainsail+Moonraker but maybe that's still worth trying out. Looks like some initial work to install missing components was already done here https://github.com/Stone-Time/Flashforge-Adventure-5M-Pro-Klipper-SSH

Either use the existing Klipper or install a clean rebuild. It looks like the printer comes with a more ore less stock Klipper 0.11, see https://github.com/FlashforgeOfficial/AD5M_Series_Klipper

b) Try to reverse engineer boot loader, run a different Linux image with Klipper etc. that is loaded from the USB Stick. I think that option takes too much effort. 

c) Run Mainsail+Moonraker on a separate device, keep (stock) Klipper as is on the Flashforge. Moonraker connects to Klipper via a Unix Domain Socket that is forwarded via Openssh to another host.

d) Only have the Klipper MCU part running on the integrated controller, Run Clipper (Python part), Moonraker, Mainsail externally.

I would probably prefer a solution that is non-invasive and does not break the built-in printer interface. Not sure about Klippers capabilities, but maybe multiple clients can connect to Klippers API server in parallel: https://www.klipper3d.org/API_Server.html

**Comment by g992** [comment]
I'm testing an invasive method right now :D 
I'm not good at programming, and uds via ssh, I think I'd be better off re-soldering a couple resistors. You could try to do a method based on software modification, that would be cool!

**Comment by xblax** [comment]
That's a cool mod, but pretty invasive indeed :D 
I will look into something software based when I have the time for it.

Did you reflash the MCU or just kept the stock firmware?

**Comment by g992** [comment]
> That's a cool mod, but pretty invasive indeed :D 
> 
> I will look into something software based when I have the time for it.
> 
> 
> 
> Did you reflash the MCU or just kept the stock firmware?

I left the stock firmware on the main board, but due to damage to the transmitters I can’t communicate with the eboard, so I’ll make hardware and software changes to it (I’ll just rebuild the klipper)

**Comment by nitronov** [comment]
> > That's a cool mod, but pretty invasive indeed :D
> > I will look into something software based when I have the time for it.
> > Did you reflash the MCU or just kept the stock firmware?
> 
> I left the stock firmware on the main board, but due to damage to the transmitters I can’t communicate with the eboard, so I’ll make hardware and software changes to it (I’ll just rebuild the klipper)

Hello! Can you describe your instructions in more detail? It is not entirely clear what you are doing in step 3, no one in the profile chats on the klipper can make out your actions. Where did you connect the soldered wires in step 2 - to the external host card uart? If so, where did you connect the soldered wires in step 3?

**Comment by g992** [comment]
> Hello! Can you describe your instructions in more detail? It is not entirely clear what you are doing in step 3, no one in the profile chats on the klipper can make out your actions. Where did you connect the soldered wires in step 2 - to the external host card uart? If so, where did you connect the soldered wires in step 3?

I don't quite understand what is not clear :(  
I used regular usb-uart converters, as in the photo below, in photos 2 and 3 of the step I showed which pins to connect to the rx and tx converter
I will redo my damaged head board in the near future and add photos to the guide
![image](https://github.com/g992/flashforge-ad5m-5mpro-research/assets/48438685/c090d6d3-5a55-417f-bafb-8639d8d1a72f)

**Comment by nitronov** [comment]
> > Hello! Can you describe your instructions in more detail? It is not entirely clear what you are doing in step 3, no one in the profile chats on the klipper can make out your actions. Where did you connect the soldered wires in step 2 - to the external host card uart? If so, where did you connect the soldered wires in step 3?
> 
> I don't quite understand what is not clear :( I used regular usb-uart converters, as in the photo below, in photos 2 and 3 of the step I showed which pins to connect to the rx and tx converter I will redo my damaged head board in the near future and add photos to the guide ![image](https://private-user-images.githubusercontent.com/48438685/302114787-c090d6d3-5a55-417f-bafb-8639d8d1a72f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwNjI5MTgsIm5iZiI6MTcwNzA2MjYxOCwicGF0aCI6Ii80ODQzODY4NS8zMDIxMTQ3ODctYzA5MGQ2ZDMtNWE1NS00MTdmLWJhZmItODYzOWQ4ZDFhNzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDE2MDMzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1NDUzNmRjMzQ3ZGQ0ZTliZTc3MDIyMzI5ZDExMzU2ZjJhZDgyMzRlYjdhY2IzOTdjYjlhYWMxZWU2Yjg1MDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PzC3l5wwchrgqoKngNqZ4LmTrlTu5xfw1N2Fa7eW6-A)

The problem is that I don't understand how to connect the uart to the board. Why did we solder the wires in step 3? What should I connect them to? In step 3, we are talking about connecting a Converter , that is, you need to connect 2 uart at once in steps 2 and 3? Or does it need to be done sequentially? My experience in printer modification is insufficient and I am grateful for the support and patience. I also want to invite you to our telegram chat on the printer https://t.me/FF_5M_5MPro I suggest we continue the conversation there. My nickname telegram is @armpix

**Comment by gpo123** [comment]
I think it is much easier to connect an external CPU to the ADM 5 via SOFTWARE.
I must admit that I don't have an ADM 5 (yet) as I am waiting for the release and reviews of the FLSUN T1.
Thanks to xblax I did a proof of concept in the meantime following his suggestion c) from above. Which means forwarding the Klipper socket on the ADM5 to an external machine.
I did this by connecting my Klipper 3D-Printer-Server with a running Klippy-instance and a Raspberry PI Zero 2W running Moonraker/Fluidd. So far the connection works fine by running:
socat "UNIX-LISTEN:/home/pi/ADM5_data/comms/klippy.sock,reuseaddr,fork" EXEC:'ssh user@PrinterServer-IP socat STDIO UNIX-CONNECT:/home/klipper/ADM_data/comms/klippy.sock'
Unfortunately this solution relies on running socat on both machines. I don't know if this command exists on the ADM5. I did not succed by trying to forward the socket by using only SSH.

**Comment by nitronov** [comment]
> I think it is much easier to connect an external CPU to the ADM 5 via SOFTWARE. I must admit that I don't have an ADM 5 (yet) as I am waiting for the release and reviews of the FLSUN T1. Thanks to xblax I did a proof of concept in the meantime following his suggestion c) from above. Which means forwarding the Klipper socket on the ADM5 to an external machine. I did this by connecting my Klipper 3D-Printer-Server with a running Klippy-instance and a Raspberry PI Zero 2W running Moonraker/Fluidd. So far the connection works fine by running: socat "UNIX-LISTEN:/home/pi/ADM5_data/comms/klippy.sock,reuseaddr,fork" EXEC:'ssh user@PrinterServer-IP socat STDIO UNIX-CONNECT:/home/klipper/ADM_data/comms/klippy.sock' Unfortunately this solution relies on running socat on both machines. I don't know if this command exists on the ADM5. I did not succed by trying to forward the socket by using only SSH.

The standard host lacks many commands, I tried to raise socat a couple of months ago, nothing worked

**Comment by gpo123** [comment]
Thats a pity, so back to SSH then. Is openssh available on the machine or is it a crippled busybox / dropbear instance ?

**Comment by xblax** [comment]
SSH is not available on the machine at all, only telnet. But I yesterday set-up a cross compilation environment so everything that's missing can be installed if needed. They use Buildroot 2019.02 with a Linaro Toolchain for the embedded OS (can look up the exact versions when I'm at home).

I already compiled and installed Dropbear because I didn't want to waste the scarce ram for Openssh.

Actually I started focusing on option a) to have Moonraker on the machine. RAM is limited, but I think it will work.  Had some troubles setting up the python venv for Moonraker, because their Python installation (3.7.11) isn't exactly standard. But I think I figured this out now and will continue exploring this in the next days.

Will be interesting be interesting to see how the stock printer application plays together with Moonraker, because they also connect to the Klipper API server via Unix Domain socket. If that makes problems they can just be cut off.  But for a general purpose mod it would for course be nice to have full stock functionality in parallel with Moonraker+Mainsail.

**Comment by nitronov** [comment]
> SSH is not available on the machine at all, only telnet. But I yesterday set-up a cross compilation environment so everything that's missing can be installed if needed. They use Buildroot 2019.02 with a Linaro Toolchain for the embedded OS (can look up the exact versions when I'm at home).
> 
> I already compiled and installed Dropbear because I didn't want to waste the scarce ram for Openssh.
> 
> Actually I started focusing on option a) to have Moonraker on the machine. RAM is limited, but I think it will work. Had some troubles setting up the python venv for Moonraker, because their Python installation (3.7.11) isn't exactly standard. But I think I figured this out now and will continue exploring this in the next days.
> 
> Will be interesting be interesting to see how the stock printer application plays together with Moonraker, because they also connect to the Klipper API server via Unix Domain socket. If that makes problems they can just be cut off. But for a general purpose mod it would for course be nice to have full stock functionality in parallel with Moonraker+Mainsail.

What if I use internal memory as a swap file?

**Comment by xblax** [comment]
> What if I use internal memory as a swap file?

Yes maybe, but I think (and hope) that's not needed.

**Comment by gpo123** [comment]
@xblax: Wow, you are making great progress !
I am astonished about SSH not being on the machine. According to stonetime's repo some sort of SSH should be on the machine (at least with the latest firmware updates ?):
https://github.com/Stone-Time/Flashforge-Adventure-5M-Pro-Klipper-SSH/blob/main/flashforge_init.sh

**Comment by gpo123** [comment]
Accidently during my fiddling with socat I connected two Moonraker instances to the same Klippy socket. That astonishingly without problems. What the Klippy socket did not like was the forwarding via OpenSSH (with my limited knowledge on this topic).

**Comment by xblax** [comment]
@gpo123 
Stonetime's repo contains a cross-built openssh. But I think that's a statically linked build. I tried to match their build enviornment so that we can link against the installed libraries.

And yes, that's a weird choice that they have telnet and not ssh by default. Instead they choose to include things like postgresql (which I think is not used). They also have two different python installations. Python 3.7.2 from Buildroot sysimage and Python 3.7.11 which they installed in /opt along with all their printer application software and a bunch of libraries. 

Their printer application is a Qt4 based touch application. So in theory someone could even develop a custom interface.

**Comment by gpo123** [comment]
Sounds great. Can you by chance provide some links to the toolchain you have used ? I am curious...

**Comment by nitronov** [comment]
> @gpo123 Stonetime's repo contains a cross-built openssh. But I think that's a statically linked build. I tried to match their build enviornment so that we can link against the installed libraries.
> 
> And yes, that's a weird choice that they have telnet and not ssh by default. Instead they choose to include things like postgresql (which I think is not used). They also have two different python installations. Python 3.7.2 from Buildroot sysimage and Python 3.7.11 which they installed in /opt along with all their printer application software and a bunch of libraries.
> 
> Their printer application is a Qt4 based touch application. So in theory someone could even develop a custom interface.

Theoretically it is possible, but is it necessary? I'm more inclined to demolish the existing system and install normal Linux

**Comment by g992** [comment]
> > @gpo123 Stonetime's repo contains a cross-built openssh. But I think that's a statically linked build. I tried to match their build enviornment so that we can link against the installed libraries.
> > And yes, that's a weird choice that they have telnet and not ssh by default. Instead they choose to include things like postgresql (which I think is not used). They also have two different python installations. Python 3.7.2 from Buildroot sysimage and Python 3.7.11 which they installed in /opt along with all their printer application software and a bunch of libraries.
> > Their printer application is a Qt4 based touch application. So in theory someone could even develop a custom interface.
> 
> Theoretically it is possible, but is it necessary? I'm more inclined to demolish the existing system and install normal Linux

Normal linux requires >128mb RAM, T113-s3 is really weak cpu, i think @xblax idea about using uds over ssh win

**Comment by nitronov** [comment]
> > > @gpo123 Stonetime's repo contains a cross-built openssh. But I think that's a statically linked build. I tried to match their build enviornment so that we can link against the installed libraries.
> > > And yes, that's a weird choice that they have telnet and not ssh by default. Instead they choose to include things like postgresql (which I think is not used). They also have two different python installations. Python 3.7.2 from Buildroot sysimage and Python 3.7.11 which they installed in /opt along with all their printer application software and a bunch of libraries.
> > > Their printer application is a Qt4 based touch application. So in theory someone could even develop a custom interface.
> > 
> > 
> > Theoretically it is possible, but is it necessary? I'm more inclined to demolish the existing system and install normal Linux
> 
> Normal linux requires >128mb RAM, T113-s3 is really weak cpu, i think @xblax idea about using uds over ssh win

More interesting is what FF wants to do in the new klipper repository. What will they provide to users?

**Comment by g992** [comment]
> > > > @gpo123 Stonetime's repo contains a cross-built openssh. But I think that's a statically linked build. I tried to match their build enviornment so that we can link against the installed libraries.
> > > > And yes, that's a weird choice that they have telnet and not ssh by default. Instead they choose to include things like postgresql (which I think is not used). They also have two different python installations. Python 3.7.2 from Buildroot sysimage and Python 3.7.11 which they installed in /opt along with all their printer application software and a bunch of libraries.
> > > > Their printer application is a Qt4 based touch application. So in theory someone could even develop a custom interface.
> > > 
> > > 
> > > Theoretically it is possible, but is it necessary? I'm more inclined to demolish the existing system and install normal Linux
> > 
> > 
> > Normal linux requires >128mb RAM, T113-s3 is really weak cpu, i think @xblax idea about using uds over ssh win
> 
> More interesting is what FF wants to do in the new klipper repository. What will they provide to users?

it`s just license agreement, klipper license require to publish klipper sources if u use it in commercial

**Comment by xblax** [comment]
> Sounds great. Can you by chance provide some links to the toolchain you have used ? I am curious...

Should be this one that I used and they also used to build the system initially:
https://releases.linaro.org/components/toolchain/binaries/7.4-2019.02/arm-linux-gnueabi/
This here probably works as well:
https://releases.linaro.org/components/toolchain/binaries/7.5-2019.12/arm-linux-gnueabi/

Newer tool-chains probably don't work unless we do static builds, because we cannot user newer glibc than 2.25 from the stock linux.

**Comment by xblax** [comment]
Short update, before anyone else runs into the same issues like I did: Python builds that are included in the system image are not usable to run Moonraker.

Buildroot System Python 3.7.2: That was compiled without of optional libraries like zlib. But zlib is required for at least pip. We can't install Moonraker dependencies.
Flashforge /opt Python 3.7.11: They compile python without _ctypes support, so that also doesn't work.

Good news: I managed to cross-compile up-to-date Python 3.11.7 for the printer. This was much easier than expected.

Next difficulty will be to install Moonraker python requirements. The dependencies contain native code which must be compiled for the target, therefore they must also be cross-built.  This should be possible using https://github.com/benfogle/crossenv

**Comment by g992** [comment]
Hi @xblax! Are there any successes? if testing is needed, maybe I can help? I was able to successfully launch a printer running a clipper with strong hardware modifications and not strong software modifications. I think I can run experiments on T113-S3 without the risk of breaking something.

**Comment by xblax** [comment]
Yes, I manged to complete proof of concept yesterday. I have Moonraker on the device and can connect with Mainsail from it! I didn't do a test print yet or anything complex. Just tried homing, movement, gcode loading, extruder and bed heat up. How well it runs is still to be determined, but I'm optimistic. It appears that Moonraker and the Stock Flashforge App can run in parallel.

I also noticed that it's not a good idea to try run Moonraker externally and just forward the Unix Socket, because it expects direct access to shared folders with Klipper (i.e. for gcode access) and Moonraker also pulls useful information from the machine like load data (RAM, CPU Usage etc.).

Next step will be to put more though in how to make a mod that others can install without breaking their printers ... I plan to do an automated installer using the stock update mechanism (copy Zip archive on USB Stick). I will share something for testing to some users once I have a first test installer. Then some can help beta testing and tuning the configurations.

I'm also currently leaning towards replacing stock Klipper as well and make the stock application connect to the custom Klipper instead, but I don't want to touch the MCU firmware. Custom Klipper will give a bit more freedom with the configuration and save some RAM, when both Klipper+Moonraker use the same python build.

**Comment by g992** [comment]
Be careful with upgrading firmware, im bricked my extruder boardwhen try to load builded klipper. I think its need more than klipper, maybe bootloader or smth.
I asked a klipper forum, but no answers at this moment.
https://github.com/g992/flashforge-ad5m-5mpro-research/issues/8

**Comment by xblax** [comment]
@g992 Yes, I don't want to flash MCU firmware. Only have the python part (klippy) and printer config separated from what is pre-installed, So that it's very easy to back to 100% stock software even if Klipper config was changed. 

Did you try to flash the stock firmware to unbrick the MCU? The flash tools and stock builds are are all included in their update packages. Do you want to make the hardware mod work long-term or rather try out the software mod.

**Comment by g992** [comment]
the main problem is that the old mcu firmware is not compatible with the new clipper :( I made the printer work, but I had to solder the controller on the printhead

**Comment by xblax** [comment]
I'm not sure if there are big benefits running newer Klipper at moment, Klipper 0.12 doesn't seem to add that much features. I'd rather stay with Klipper 0.11~ for now that is compatible with stock MCU firmware.

**Comment by jlodew0** [comment]
the t113-S3 could be upgraded with the S4 (256MB version - same die, so pin compatible).
https://nl.aliexpress.com/w/wholesale-t113%2525252ds4.html?spm=a2g0o.productlist.search.0
Allwinner kernel available here: https://github.com/Tina-Linux/linux-5.4
I only did embedded projects with u-boot in the past. Seems like only awboot and xboot are working atm acording to this page:
https://linux-sunxi.org/T113-s3

p.s. using an external host like rpi instead of the onboard soc would render the display useless as well as the webcam on the pro machines.

p.s. 2 does anyone knows how many and what type of load cells are used on the bed for leveling?
And how about klipper support because I can only find some topics about some users writing stuff for load cells in klipper. If you want to make a firmware for the mcu you might want to overcome that part + some custom things implemented as well. I didn't went trough the config file yet, but it wouldn't surprise me that isn't all standard klipper stuff in there.

I'm afraid that since FF is yet another chinese company that they don't care a lot about licensing. So opening up everything as required with the gpl license is probably not going to happen. However, I hope to be proven wrong.
their current github repo seems to be setup just to comply to gpl (and even those requirements aren't fully met)

They have a good machine hardware wise and price wise, it beats the bambu products by a large margin.
If they would only see that opening up the machine's source code would increase their sales by a huge factor (that's what I would expect, at least by the more advanced maker/thinker community).

**Comment by g992** [comment]
I’m not sure that the upgrade to 256MB of RAM will be available to all users, as it requires soldering skills + skills in working with embedded Linux. Regarding the table level cells, judging by the klipper there are 3 of them and apparently there is an additional controller that processes them, since the clipper receives the “probe” signal from one pin. with an external controller it works the same way, I was able to get a height map. Regarding the display: I plan to replace it with a 4.3-inch display with hdmi, but these are just plans. As an experiment, I want to replace almost all the electronics of this printer.

**Comment by xblax** [comment]
I don't think that the 128MB of RAM are a deal-breaker. We just have to be a bit resourceful with what we install, but Klipper+Moonraker+Mainsail/Fluid should just run fine. I'm definitely not going to solder another SOC on the board, the amount of work required for that is just not worth it. Processing power also seems to be enough.

What I'm currently working on is a change root environment based on Buildroot 2023.02 that basically runs everything but the stock printer app. In the chroot environment we can have a custom klipper (klipppy), moonraker and a webserver to provide mainsail & fluidd (probably busybox-httpd). Basically a custom linux in linux. Dropbear with root access will also run in the chroot, so we don't even have to edit the root password. I've started a git project for this, but that's not yet public.

I everything goes as planned, the mod installs just an Init script and a self-contained folder on the /data partition with the mod system. Going back to stock functionality is than as easy as just deleting one folder or the Init script.

**Comment by g992** [comment]
Guys says that mellow have pi lite board on t113-s3, that can run klipper + mooonraker + web interface. Some docs here: https://mellow.klipper.cn/#/introduction/downloadimg
Maybe it can help

**Comment by xblax** [comment]
So I managed to do print a test Benchy with Moonraker, Klipper, Mainsail running on the machine yesterday. This worked quite well. The Ram is sufficient, CPU Usage never went above 20%. The difficulty is how to set up everything properly.

The Stock app does not really like it to not be in control of Klipper. I think it's better to disable it while using Moonraker, otherwise this can lead to unexpected issues. I think the best for Klipper Mod mode would be to just copy the Wifi settings from the stock app and enable DHCP for Ethernet - for Filament loading etc. we then need Klipper Macros.

I also tried running slightly newer Klipper than 0.11, but that didn't work well with the MCU firmware. I think they need to be reflashed for newer Klipper.

![Screenshot_2024-02-10_20-56-05](https://github.com/g992/flashforge-ad5m-5mpro-research/assets/9802544/7d250242-665e-4acf-bb35-566ce0dfb7f6)
![Screenshot_2024-02-10_21-02-26](https://github.com/g992/flashforge-ad5m-5mpro-research/assets/9802544/45425f1e-8a57-4bf3-bbba-8e17a1a4a4ac)
![Screenshot_2024-02-10_21-01-10](https://github.com/g992/flashforge-ad5m-5mpro-research/assets/9802544/64ddf9a9-162e-4901-a1cb-fb2f2164e3e7)

**Comment by g992** [comment]
Nice! I think the upgrade to 0.12 is not so important, since it turned out to be more difficult to assemble the firmware on the N32 than we thought. I think the stable operation of the 0.11 version is enough
we have a community of owners who want to install a moonraker, could you share the basic steps? at least not through the update files, but through the terminal

**Comment by xblax** [comment]
Yes ... hope I have soon something that I can share. The problem is that build instructions are too complicated to write them down in a tutorial or something like that. I started to script most things, fortunately.

The whole thing works in a chroot environment, which is cross-build with buildroot. There are a lot of step involved in building all dependencies, unfortunately. 

I will try to get it in a state that it can install and run (maybe manual startup via telnet at first).

**Comment by nitronov** [comment]
> Yes ... hope I have soon something that I can share. The problem is that build instructions are too complicated to write them down in a tutorial or something like that. I started to script most things, fortunately.
> 
> The whole thing works in a chroot environment, which is cross-build with buildroot. There are a lot of step involved in building all dependencies, unfortunately.
> 
> I will try to get it in a state that it can install and run (maybe manual startup via telnet at first).

As an option to record a video, I think anyone can figure it out, and if not, then they won’t need it

**Comment by xblax** [comment]
> As an option to record a video, I think anyone can figure it out, and if not, then they won’t need it

That would have to be a multiple hour session :D
It's easier for me to share something prebuilt for testing and then make the git repos with build scripts public, eventually.

Currently I'm having a problem with the MCUs. Klipper cannot connect after Linux startup. Only if the Flashforge software starts with their Klipper, then I kill that, and then my Klipper can connect. I don't think it's a change to Klipper they did, I looked at https://github.com/FlashforgeOfficial/AD5M_Series_Klipper - that is basically vanilla Klipper.

There must be some initialization missing. Do we have to power on the MCUs maybe?
@g992 Any Idea - you looked at the MCUs more than I did.

**Comment by nitronov** [comment]
> > As an option to record a video, I think anyone can figure it out, and if not, then they won’t need it
> 
> That would have to be a multiple hour session :D It's easier for me to share something prebuilt for testing and then make the git repos with build scripts public, eventually.
> 
> Currently I'm having a problem with the MCUs. Klipper cannot connect after Linux startup. Only if the Flashforge software starts with their Klipper, then I kill that, and then my Klipper can connect. I don't think it's a change to Klipper they did, I looked at https://github.com/FlashforgeOfficial/AD5M_Series_Klipper - that is basically vanilla Klipper.
> 
> There must be some initialization missing. Do we have to power on the MCUs maybe? @g992 Any Idea - you looked at the MCUs more than I did.

Half of the system is launched through the installed proprietary Flashforge software, the same display is not registered in the configuration settings, apparently it is also launched through native software

**Comment by g992** [comment]
> > As an option to record a video, I think anyone can figure it out, and if not, then they won’t need it
> 
> That would have to be a multiple hour session :D It's easier for me to share something prebuilt for testing and then make the git repos with build scripts public, eventually.
> 
> Currently I'm having a problem with the MCUs. Klipper cannot connect after Linux startup. Only if the Flashforge software starts with their Klipper, then I kill that, and then my Klipper can connect. I don't think it's a change to Klipper they did, I looked at https://github.com/FlashforgeOfficial/AD5M_Series_Klipper - that is basically vanilla Klipper.
> 
> There must be some initialization missing. Do we have to power on the MCUs maybe? @g992 Any Idea - you looked at the MCUs more than I did.

hmm, how ff software starting up? (I found many strange functionals like pin enabling (?)) if FF software starts from init.d, maybe edit start script to kill software after 10s from launch? :D
I killed the motherboard, so I can safely explore it. I will deal with this in the coming days, perhaps it will be necessary to add a simple script to enable the MCU

**Comment by xblax** [comment]
> Half of the system is launched through the installed proprietary Flashforge software, the same display is not registered in the configuration settings, apparently it is also launched through native software

Yes, some things are ... but display is available at /dev/fb0 - I didn't figure out the framebuffer format and how to display images at it. But installer just writes framebuffer images to /dev/fb0 

> hmm, how ff software starting up? (I found many strange functionals like pin enabling (?)) if FF software starts from init.d, maybe edit start script to kill software after 10s from launch? :D

Did you disassemble the software or where did you find it? Maybe that would help me.

I don't want to start flashforge software. I want to hijack system init before flashforge software launches and continue in chroot environment.

**Comment by g992** [comment]
> I don't want to start flashforge software. I want to hijack system init before flashforge software launches and continue in chroot environment.

I disassembled it, but I didn't figure it out much
, I found a lot of tracks that control something and come from T113, as well as update files depend on the system gpios (try running NationsCommand on RP)
I will investigate them, write in this topic what I find

**Comment by nitronov** [comment]
> > Half of the system is launched through the installed proprietary Flashforge software, the same display is not registered in the configuration settings, apparently it is also launched through native software
> 
> Yes, some things are ... but display is available at /dev/fb0 - I didn't figure out the framebuffer format and how to display images at it. But installer just writes framebuffer images to /dev/fb0
> 
> > hmm, how ff software starting up? (I found many strange functionals like pin enabling (?)) if FF software starts from init.d, maybe edit start script to kill software after 10s from launch? :D
> 
> Did you disassemble the software or where did you find it? Maybe that would help me.
> 
> I don't want to start flashforge software. I want to hijack system init before flashforge software launches and continue in chroot environment.

I'm now experimenting with loading a different loader. All that remains is to find out the address of the cell where the u-boot is installed. Any ideas on how to find out the address?

**Comment by g992** [comment]
@xblax 
Do you speak Russian by any chance? (in places, the translation was "clipper", I have this happening when I use a translator)

**Comment by xblax** [comment]
@g992 
No, not at all :D 

Here I have the GPIO state before Flashforge software starts. But this output is unchanged after Flashforge software start. I will try to use `gpiomon`. 
```
ADM5 ~ # gpioinfo gpiochip0
gpiochip0 - 224 lines:
	line   0:      unnamed       unused   input  active-high 
	line   1:      unnamed       unused   input  active-high 
	line   2:      unnamed       unused   input  active-high 
	line   3:      unnamed       unused   input  active-high 
	line   4:      unnamed       unused   input  active-high 
	line   5:      unnamed       unused   input  active-high 
	line   6:      unnamed       unused   input  active-high 
	line   7:      unnamed       unused   input  active-high 
	line   8:      unnamed       unused   input  active-high 
	line   9:      unnamed       unused   input  active-high 
	line  10:      unnamed       unused   input  active-high 
	line  11:      unnamed       unused   input  active-high 
	line  12:      unnamed       unused   input  active-high 
	line  13:      unnamed       unused   input  active-high 
	line  14:      unnamed       unused   input  active-high 
	line  15:      unnamed       unused   input  active-high 
	line  16:      unnamed       unused   input  active-high 
	line  17:      unnamed       unused   input  active-high 
	line  18:      unnamed       unused   input  active-high 
	line  19:      unnamed       unused   input  active-high 
	line  20:      unnamed       unused   input  active-high 
	line  21:      unnamed       unused   input  active-high 
	line  22:      unnamed       unused   input  active-high 
	line  23:      unnamed       unused   input  active-high 
	line  24:      unnamed       unused   input  active-high 
	line  25:      unnamed       unused   input  active-high 
	line  26:      unnamed       unused   input  active-high 
	line  27:      unnamed       unused   input  active-high 
	line  28:      unnamed       unused   input  active-high 
	line  29:      unnamed       unused   input  active-high 
	line  30:      unnamed       unused   input  active-high 
	line  31:      unnamed       unused   input  active-high 
	line  32:      unnamed       unused   input  active-high 
	line  33:      unnamed       kernel   input  active-high [used]
	line  34:      unnamed       unused   input  active-high 
	line  35:      unnamed  "usb1-vbus"  output  active-high [used]
	line  36:      unnamed       unused  output  active-high 
	line  37:      unnamed       unused  output  active-high 
	line  38:      unnamed       kernel   input  active-high [used]
	line  39:      unnamed       kernel   input  active-high [used]
	line  40:      unnamed       unused   input  active-high 
	line  41:      unnamed       unused   input  active-high 
	line  42:      unnamed       unused   input  active-high 
	line  43:      unnamed       unused   input  active-high 
	line  44:      unnamed       unused   input  active-high 
	line  45:      unnamed       unused   input  active-high 
	line  46:      unnamed       unused   input  active-high 
	line  47:      unnamed       unused   input  active-high 
	line  48:      unnamed       unused   input  active-high 
	line  49:      unnamed       unused   input  active-high 
	line  50:      unnamed       unused   input  active-high 
	line  51:      unnamed       unused   input  active-high 
	line  52:      unnamed       unused   input  active-high 
	line  53:      unnamed       unused   input  active-high 
	line  54:      unnamed       unused   input  active-high 
	line  55:      unnamed       unused   input  active-high 
	line  56:      unnamed       unused   input  active-high 
	line  57:      unnamed       unused   input  active-high 
	line  58:      unnamed       unused   input  active-high 
	line  59:      unnamed       unused   input  active-high 
	line  60:      unnamed       unused   input  active-high 
	line  61:      unnamed       unused   input  active-high 
	line  62:      unnamed       unused   input  active-high 
	line  63:      unnamed       unused   input  active-high 
	line  64:      unnamed       unused   input  active-high 
	line  65:      unnamed       unused   input  active-high 
	line  66:      unnamed       kernel   input  active-high [used]
	line  67:      unnamed       kernel   input  active-high [used]
	line  68:      unnamed       kernel   input  active-high [used]
	line  69:      unnamed       kernel   input  active-high [used]
	line  70:      unnamed       kernel   input  active-high [used]
	line  71:      unnamed       kernel   input  active-high [used]
	line  72:      unnamed       unused   input  active-high 
	line  73:      unnamed       unused   input  active-high 
	line  74:      unnamed       unused   input  active-high 
	line  75:      unnamed       unused   input  active-high 
	line  76:      unnamed       unused   input  active-high 
	line  77:      unnamed       unused   input  active-high 
	line  78:      unnamed       unused   input  active-high 
	line  79:      unnamed       unused   input  active-high 
	line  80:      unnamed       unused   input  active-high 
	line  81:      unnamed       unused   input  active-high 
	line  82:      unnamed       unused   input  active-high 
	line  83:      unnamed       unused   input  active-high 
	line  84:      unnamed       unused   input  active-high 
	line  85:      unnamed       unused   input  active-high 
	line  86:      unnamed       unused   input  active-high 
	line  87:      unnamed       unused   input  active-high 
	line  88:      unnamed       unused   input  active-high 
	line  89:      unnamed       unused   input  active-high 
	line  90:      unnamed       unused   input  active-high 
	line  91:      unnamed       unused   input  active-high 
	line  92:      unnamed       unused   input  active-high 
	line  93:      unnamed       unused   input  active-high 
	line  94:      unnamed       unused   input  active-high 
	line  95:      unnamed       unused   input  active-high 
	line  96:      unnamed       unused   input  active-high 
	line  97:      unnamed       unused   input  active-high 
	line  98:      unnamed       unused   input  active-high 
	line  99:      unnamed       unused   input  active-high 
	line 100:      unnamed       unused   input  active-high 
	line 101:      unnamed       unused   input  active-high 
	line 102:      unnamed       unused   input  active-high 
	line 103:      unnamed       unused   input  active-high 
	line 104:      unnamed       unused   input  active-high 
	line 105:      unnamed       unused   input  active-high 
	line 106:      unnamed       kernel   input  active-high [used]
	line 107:      unnamed       kernel   input  active-high [used]
	line 108:      unnamed       kernel   input  active-high [used]
	line 109:      unnamed       kernel   input  active-high [used]
	line 110:      unnamed       unused   input  active-high 
	line 111:      unnamed       unused   input  active-high 
	line 112:      unnamed       unused   input  active-high 
	line 113:      unnamed       unused   input  active-high 
	line 114:      unnamed       unused   input  active-high 
	line 115:      unnamed       unused   input  active-high 
	line 116:      unnamed       unused   input  active-high 
	line 117:      unnamed       unused   input  active-high 
	line 118:      unnamed       unused   input  active-high 
	line 119:      unnamed       unused   input  active-high 
	line 120:      unnamed       unused   input  active-high 
	line 121:      unnamed       unused   input  active-high 
	line 122:      unnamed       unused   input  active-high 
	line 123:      unnamed       unused   input  active-high 
	line 124:      unnamed       unused   input  active-high 
	line 125:      unnamed       unused   input  active-high 
	line 126:      unnamed       unused   input  active-high 
	line 127:      unnamed       unused   input  active-high 
	line 128:      unnamed       kernel   input  active-high [used]
	line 129:      unnamed       kernel   input  active-high [used]
	line 130:      unnamed       kernel   input  active-high [used]
	line 131:      unnamed       kernel   input  active-high [used]
	line 132:      unnamed       kernel   input  active-high [used]
	line 133:      unnamed       kernel   input  active-high [used]
	line 134:      unnamed       kernel   input  active-high [used]
	line 135:      unnamed       kernel   input  active-high [used]
	line 136:      unnamed       kernel   input  active-high [used]
	line 137:      unnamed       kernel   input  active-high [used]
	line 138:      unnamed       kernel   input  active-high [used]
	line 139:      unnamed       unused   input  active-high 
	line 140:      unnamed       unused   input  active-high 
	line 141:      unnamed    "phy-rst"  output  active-high [used]
	line 142:      unnamed       unused   input  active-high 
	line 143:      unnamed       unused   input  active-high 
	line 144:      unnamed       unused   input  active-high 
	line 145:      unnamed       unused   input  active-high 
	line 146:      unnamed       unused   input  active-high 
	line 147:      unnamed       unused   input  active-high 
	line 148:      unnamed       unused   input  active-high 
	line 149:      unnamed       unused   input  active-high 
	line 150:      unnamed       unused   input  active-high 
	line 151:      unnamed       unused   input  active-high 
	line 152:      unnamed       unused   input  active-high 
	line 153:      unnamed       unused   input  active-high 
	line 154:      unnamed       unused   input  active-high 
	line 155:      unnamed       unused   input  active-high 
	line 156:      unnamed       unused   input  active-high 
	line 157:      unnamed       unused   input  active-high 
	line 158:      unnamed       unused   input  active-high 
	line 159:      unnamed       unused   input  active-high 
	line 160:      unnamed       kernel   input  active-high [used]
	line 161:      unnamed       kernel   input  active-high [used]
	line 162:      unnamed       kernel   input  active-high [used]
	line 163:      unnamed       kernel   input  active-high [used]
	line 164:      unnamed       kernel   input  active-high [used]
	line 165:      unnamed       kernel   input  active-high [used]
	line 166:      unnamed       unused   input  active-high 
	line 167:      unnamed       unused   input  active-high 
	line 168:      unnamed       unused   input  active-high 
	line 169:      unnamed       unused   input  active-high 
	line 170:      unnamed       unused   input  active-high 
	line 171:      unnamed       unused   input  active-high 
	line 172:      unnamed       unused   input  active-high 
	line 173:      unnamed       unused   input  active-high 
	line 174:      unnamed       unused   input  active-high 
	line 175:      unnamed       unused   input  active-high 
	line 176:      unnamed       unused   input  active-high 
	line 177:      unnamed       unused   input  active-high 
	line 178:      unnamed       unused   input  active-high 
	line 179:      unnamed       unused   input  active-high 
	line 180:      unnamed       unused   input  active-high 
	line 181:      unnamed       unused   input  active-high 
	line 182:      unnamed       unused   input  active-high 
	line 183:      unnamed       unused   input  active-high 
	line 184:      unnamed       unused   input  active-high 
	line 185:      unnamed       unused   input  active-high 
	line 186:      unnamed       unused   input  active-high 
	line 187:      unnamed       unused   input  active-high 
	line 188:      unnamed       unused   input  active-high 
	line 189:      unnamed       unused   input  active-high 
	line 190:      unnamed       unused   input  active-high 
	line 191:      unnamed       unused   input  active-high 
	line 192:      unnamed       unused   input  active-high 
	line 193:      unnamed       unused   input  active-high 
	line 194:      unnamed       unused   input  active-high 
	line 195:      unnamed       unused   input  active-high 
	line 196:      unnamed       kernel   input  active-high [used]
	line 197:      unnamed       kernel   input  active-high [used]
	line 198:      unnamed       kernel   input  active-high [used]
	line 199:      unnamed       kernel   input  active-high [used]
	line 200:      unnamed       kernel   input  active-high [used]
	line 201:      unnamed       kernel   input  active-high [used]
	line 202:      unnamed       unused   input  active-high 
	line 203:      unnamed       unused   input  active-high 
	line 204:      unnamed       unused   input  active-high 
	line 205:      unnamed       unused   input  active-high 
	line 206:      unnamed       unused   input  active-high 
	line 207:      unnamed       unused   input  active-high 
	line 208:      unnamed       unused   input  active-high 
	line 209:      unnamed       unused   input  active-high 
	line 210:      unnamed       unused   input  active-high 
	line 211:      unnamed       unused   input  active-high 
	line 212:      unnamed       unused   input  active-high 
	line 213:      unnamed       unused   input  active-high 
	line 214:      unnamed       unused   input  active-high 
	line 215:      unnamed       unused   input  active-high 
	line 216:      unnamed       unused   input  active-high 
	line 217:      unnamed       unused   input  active-high 
	line 218:      unnamed       unused   input  active-high 
	line 219:      unnamed       unused   input  active-high 
	line 220:      unnamed       unused   input  active-high 
	line 221:      unnamed       unused   input  active-high 
	line 222:      unnamed       unused   input  active-high 
	line 223:      unnamed       unused   input  active-high
```

**Comment by g992** [comment]
hmm, any errors in moonrakers logs? 
and usd alive when system starts? maybe ff software creating uds (?)
or maybe it starts klipper? i didnt remember klipper at init.d

**Comment by xblax** [comment]
> hmm, any errors in moonrakers logs? and usd alive when system starts? maybe ff software creating uds (?) or maybe it starts klipper? i didnt remember klipper at init.d

No, Klipper was the issue, could not connect to eboard-MCU. Yes, normally FF-software starts Klipper, but I don't start FF-software for the mod.

But it's working now, cannot reproduce it. No GPIOs needed. Don't know what happended earlier, maybe MCU shut down for some reason.

**Comment by g992** [comment]
I got the same bug (klipper doesn’t see the main controller, but my uart is hardwired to raspberry) and... it went away after 2-3 reboots! and now I can’t reproduce it either! I don’t know where it comes from, and how FF fights it, maybe there is a gpio for rebooting the controller, I need to check it

**Comment by g992** [comment]
I got the same bug (klipper doesn’t see the main controller, but my uart is hardwired to raspberry) and... it went away after 2-3 reboots! and now I can’t reproduce it either! I don’t know where it comes from, and how FF fights it, maybe there is a gpio for rebooting the controller, I need to check it

**Comment by xblax** [comment]
Also got the bug again yesterday. Only way to resolve it was to boot into stock mode, FF-Application didn't have problems, then reboot into Mod mode. Maybe the MCU must be enabled once after cold boot.

I also don't really know yet what some of the GPIO Pins and Sensors in the Klipper config do.

**Comment by xblax** [comment]
Ok so I started to look at a bit at the "firmwareExe" from Flashforge. The pattern seems to be, that after power up, the software must run or otherwise the eboard-MCU is not available. The main-MCU is ready all the time without additional initialization. 

First I had a look at the output on console. Logs saved at `/mnt/data/logFiles/firmwareExe.log` don't help much. MCU initialization happens earlier. But if we launch it on the console, we see some serial initialization is taking place:

```
Serial: serial fd: 14
Serial: serial ready recv:  , count: 0
Serial: serial ready recv: Ready.
 , count: 8
Serial: serial ready end: 1
Serial: serial ACK send: A , count: 1
Serial: serial ACK recv: 6 , count: 1
```

The sequence is a bit different every time, sometimes it's quite a bit longer with more commands exchanged. I looked into the binary with Ghidra and found the relevant part. There are two functions `bootSerialMainMcu` and `bootSerialMainEboardMcu`. The one for the main MCU is never called, that matches obervations that it doesn't need initialization. Maybe they used it during development. But `bootSerialMainEboardMcu` is called directly from main().

Here are the code snippets from Ghidra: https://gist.github.com/xblax/2287bc83f399d2144952cd345518cf78

I assume that MCU boots in program or debug mode and must be told to start the actual program first. I guess this could be reverse engineered, but do we maybe have a datasheet that explains this?

**Comment by xblax** [comment]
That unsolved problem didn't let me rest so I had to reverse engineer that code ... seems they have written a custom bootloader. I had checked STM and Nations Bootloader descriptions, but that didn't look like what was sent.

What the Eboard-MCU does after start, is to periodically send:
```
Ready.
Ready.
Ready.
Ready.
...
```
That is detected by the initialization code and with command 'A' (0x41) the boot loader starts the application. This command is acknowledged by the bootloader with byte 0x06 (Ascii ACK). I have updated the Gist with C-Code to do that initialization:
https://gist.github.com/xblax/2287bc83f399d2144952cd345518cf78

Now I can start Klipper, Moonraker etc. after boot 😎 
Will try to get something ready for testing for you soon.

**Comment by nitronov** [comment]
> Эта нерешенная проблема не давала мне покоя, поэтому мне пришлось перепроектировать этот код... кажется, они написали собственный загрузчик. Я проверил описания загрузчика STM и Nations, но они не были похожи на то, что было отправлено.
> 
> После запуска Eboard-MCU периодически отправляет:
> 
> ```
> Ready.
> Ready.
> Ready.
> Ready.
> ...
> ```
> 
> Это определяется кодом инициализации, и с помощью команды «A» (0x41) загрузчик запускает приложение. Эта команда подтверждается загрузчиком с помощью байта 0x06 (Ascii ACK). Я обновил Gist с помощью C-кода для выполнения этой инициализации: [https://gist.github.com/xblax/2287bc83f399d2144952cd345518cf78 ](https://gist.github.com/xblax/2287bc83f399d2144952cd345518cf78)
> 
> Теперь я могу запустить Klipper, Moonraker и т. д. после загрузки 😎 Постараюсь в ближайшее время подготовить для вас что-нибудь для тестирования.

I'm still trying to flash another bootloader, I can't find the address of the cell where the u-boot is located to erase it

**Comment by xblax** [comment]
@nitronov 

> I'm still trying to flash another bootloader, I can't find the address of the cell where the u-boot is located to erase it

Do know if the device is using u-boot at all? I didn't check it, but @jlodew0 wrote earlier that T113-S3 doesn't support u-boot:
>  I only did embedded projects with u-boot in the past. Seems like only awboot and xboot are working atm acording to this page:
https://linux-sunxi.org/T113-s3

**Comment by nitronov** [comment]
> @nitronov
> 
> > I'm still trying to flash another bootloader, I can't find the address of the cell where the u-boot is located to erase it
> 
> Do know if the device is using u-boot at all? I didn't check it, but @jlodew0 wrote earlier that T113-S3 doesn't support u-boot:
> 
> > I only did embedded projects with u-boot in the past. Seems like only awboot and xboot are working atm acording to this page:
> > https://linux-sunxi.org/T113-s3

U-boot version 2018.05 installed

**Comment by g992** [comment]
> That unsolved problem didn't let me rest so I had to reverse engineer that code ... seems they have written a custom bootloader. I had checked STM and Nations Bootloader descriptions, but that didn't look like what was sent.
> 
> What the Eboard-MCU does after start, is to periodically send:
> 
> ```
> Ready.
> Ready.
> Ready.
> Ready.
> ...
> ```
> 
> That is detected by the initialization code and with command 'A' (0x41) the boot loader starts the application. This command is acknowledged by the bootloader with byte 0x06 (Ascii ACK). I have updated the Gist with C-Code to do that initialization: https://gist.github.com/xblax/2287bc83f399d2144952cd345518cf78
> 
> Now I can start Klipper, Moonraker etc. after boot 😎 Will try to get something ready for testing for you soon.

Niiice, better, man

**Comment by jlodew0** [comment]
@nitronov and @xblax 

I didn't have any time on spare yet to thinker with my 5m yet.
So I'm just refering to what I find online. 
Support for u-boot was being worked on end 2022/beginning of 2023 --> https://lore.kernel.org/u-boot/20221206004549.29015-1-andre.przywara@arm.com/
Anyway, why would you want to get rid of u-boot? 

p.s. @g992 do you use a rpi 4?
Because I read that you used ch340 based uart's, rpi has 6 uart's (5 usable because uart 0 and 1 shares the same gpio's).
https://raspberrypi.stackexchange.com/questions/104464/where-are-the-uarts-on-the-raspberry-pi-4

I think I'm going the rpi route as well. That way, I don't have to worry about limited resources.

**Comment by xblax** [comment]
@g992 @nitronov @gpo123 @jlodew0 
I've now manged to put a first alpha release together for my mod. The repository is still private, but I've added you as collaborators for now. Please don't commit anything there, I think that is just the easiest way do a first release with a closed group of testers. If some can confirm that it works - especially also on the 5M Pro, then I will make the repo fully public.

https://github.com/xblax/flashforge_adm5_klipper_mod
Here is a public copy of the README.md for everyone else who is interested:
https://gist.github.com/xblax/81fbb8c5671c206276f9f6070ef0cafa

If someone else want's to try this out already, maybe leave a comment here. The mod has no WiFi yet, only Ethernet LAN. Probably that's a deal-breaker for some. 

>I think I'm going the rpi route as well. That way, I don't have to worry about limited resources.

It really seems to work fine for me. The mod has no camera integration yet so that will add a bit to resource usage, but I've ordered the camera kit for the 5M and will try this out.

**Comment by AJolly** [comment]
Add me, I should get mine in a week.

**Comment by KaruroChori** [comment]
That's great news! If you need additional testers I can do it.
I have a 5m with root access, already updated to the latest firmware and working via ethernet, and I cannot wait to stop using that UI on the small touchscreen :D.

**Comment by ellaleks** [comment]
Я тоже хочу протестировать. В наличии Adventurer 5M. Буду благодарен за доверие.

**Comment by xblax** [comment]
@AJolly @KaruroChori @ellaleks 
Added you as well as collaborators to the repo.

**Comment by NokHook** [comment]
> Please

Also wants to test. I have AD5M)

**Comment by Moonfizzle** [comment]
Happy to do some testing also. I have a rooted 5M and a camera installed

**Comment by xblax** [comment]
@NokHook @Moonfizzle 
Added you.

I've also enabled the Discussions section in the repo: https://github.com/xblax/flashforge_adm5_klipper_mod/discussions
That can be used for discussing the mod. Or open an issue in the mod repo if something that should work does not.

I'm not sure if @g992 is happy if we completely spam this issue here with comments.

@Moonfizzle 
mjpg-streamer is available in the mod, you can try to start it manually via SSH. But maybe some dependencies for camera support are still missing.

**Comment by x2o** [comment]
@xblax I'd be happy to test as well if you need more people on the Alpha release, rooted AD5M

**Comment by stenny92** [comment]
@xblax I just started following this thread. I would like to be a tester for this.

**Comment by mpolr** [comment]
@xblax pls add me, I have adm5 with camera

**Comment by consp** [comment]
@xblax keep up the good work!

It's a bit of a shame though this thread is now hijacked to discuss the mod, so a minor suggestion: You can also create two repo's and lock it down as needed by keeping code in the private and readme/info in the public repository (create two remotes locally, only push the public branch to the public one). That way the issue and discussion systems are available for the world to see and you can sync the branches if the time arises.

My printer will only arrive by the weekend but it will definitely be open before the calibration has finished. (no need to add me)

**Comment by xblax** [comment]
@x2o  @stenny92  @mpolr 
Added you. Please check discussions and open issues in repo before testing. There are some improvements in the printer config not yet in the alpha build. You can update the configs manually in mainsail. Currently working on the bed leveling routine. Printer profile in Slicer software should be adjusted for Klipper before trying a print.

Will now wait a bit before adding more for testing.

**Comment by xblax** [comment]
@consp 
I think it's too late to split this up now, but mod related discussions and issues around it should be kept together long-term anyway. I don't want to keep this closed for too long, but also don't want to publish anything where I assume people could accidentally break something or get the printer into a non-functional state that is hard to resolve. 

I just accidentally messed up the load cell calibration for bed leveling by setting an output pin that is in the original Klipper config. Nozzle kept crashing into the bed when attempting bed leveling after that, even with stock software. Had to reverse engineer the original software software and find a hidden calibration menu to fix this. Original Flashforge software does much trough G-Code commands sent to Klipper, i.e. setting certain output pins before bed leveling. But I'm getting an understand how this works, bit by bit.

Feel free to continue technical discussion around Klipper here as well. I intended this as an open-ended discussion and my mod is just what came out of it so far.

**Comment by khayrulinnikita** [comment]
@xblax, please add me

**Comment by consp** [comment]
> I think it's too late to split this up now, but mod related discussions and issues around it should be kept together long-term anyway. I don't want to keep this closed for too long, but also don't want to publish anything where I assume people could accidentally break something or get the printer into a non-functional state that is hard to resolve.

Makes sense. Though when I do that I usually add a point like "now use <programmer x> to flash the data" which scares off people who do not know what they are doing.

> I just accidentally messed up the load cell calibration for bed leveling by setting an output pin that is in the original Klipper config. Nozzle kept crashing into the bed when attempting bed leveling after that, even with stock software. Had to reverse engineer the original software software and find a hidden calibration menu to fix this. Original Flashforge software does much trough G-Code commands sent to Klipper, i.e. setting certain output pins before bed leveling. But I'm getting an understand how this works, bit by bit.

Interesting as well, sounds like a hack, but hacks you can use in your own code. You definitely do not want to screw load cell calibration if you can. For those testing: Get a mains power switch nearby so you have an emergency power-off in case of head crashes.

@xblax 
Do you know if the "debug" pins on the board are ttyS3 which is the serial console (from memory so actual tty might be off)? Since if it is you can reuse it for other serial communication.

**Comment by xblax** [comment]
> Interesting as well, sounds like a hack, but hacks you can use in your own code. You definitely do not want to screw load cell calibration if you can. For those testing: Get a mains power switch nearby so you have an emergency power-off in case of head crashes.

Yes. I think load cell calibration with a reference weight is done once at factory. Then for leveling we must only set the load cell tare after heating the bed. But at least I do now know how that works.

> Do you know if the "debug" pins on the board are ttyS3 which is the serial console (from memory so actual tty might be off)? Since if it is you can reuse it for other serial communication.

No idea. I only know that it has a debug serial console, but never used it myself https://www.reddit.com/r/FlashForge/comments/1ahwari/shell_access_password_for_flashforge_printers/kordy9a/

But there are plenty of USB ports available according to the excellent hardware documentation already done by @g992 
You can attach an USB to Serial converter there if needed.

**Comment by g992** [comment]
> Do you know if the "debug" pins on the board are ttyS3 which is the serial console (from memory so actual tty might be off)? Since if it is you can reuse it for other serial communication.

the debug port on the board is just an access to the linux terminal, I connected to it via a usb serial adapter and got a full-fledged terminal. it can be used to access the terminal, but I do not know why it is when there is access via telnet/ssh
. Once it was useful to me when I clogged the file system and telnet/ssh simply could not start, then I deleted the extra files through this terminal

**Comment by consp** [comment]
> the debug port on the board is just an access to the linux terminal

I guessed as much from the firmware images, ttyS3 is bound to it.

>  Once it was useful to me when I clogged the file system and telnet/ssh simply could not start, then I deleted the extra files through this terminal

I'm pretty sure that is the reason, it's labled debug. I might also be handy if uboot was used (or any other bootloader). Mine will arrive in about 2 hours so fingers crossed.

**Comment by Ross922** [comment]
Forgive my ignorance but do you think its possible a company like BTT will come out with a new MB for the 5M Pro if it gains enough traction? That way the weak cpu and MB package can be bypassed etc? Note: I am not familiar with what it takes to put klipper onto this machine nor root and mod for klipper to work on the stock HW, im just asking what is likely an uninformed general question :-)

**Comment by g992** [comment]
> Forgive my ignorance but do you think its possible a company like BTT will come out with a new MB for the 5M Pro if it gains enough traction? That way the weak cpu and MB package can be bypassed etc? Note: I am not familiar with what it takes to put klipper onto this machine nor root and mod for klipper to work on the stock HW, im just asking what is likely an uninformed general question :-) 

I don’t think so, only if in an official collab. The stock CPU is enough to run clipper + moonraker + mainsale + fluidd. There are also hopes that it will be enough to run clipperscreen, then we will get a full-fledged printer

**Comment by consp** [comment]
> also hopes that it will be enough to run clipperscreen

Running X will be a stretch, klipperscreen is designed for hardware an order of magnitude faster (ok, not that much but you get the idea). Anything which uses a bitmapped framebuffer might work. Afaik the application (firmwareExe) is a Qt4.8.6 applet outputting to the fb with the qt embedded display driver, from what I can tell with a quick look.

The device is a simple 800x960 pixel framebuffer so no technical limitations there. Should be doable (at least if gives nice noise when doing `dd if=/dev/urandom bs=1 count=3072000 > /dev/fb0`)

I already had the original FW crash on me once so hopefully you can get it working.

Other thing of note I came across: the power switch is a file. I'll test it after I finished printing.

```
[INFO]1970/01/01 00:00:15 set QT_QPA_PLATFORM: linuxfb:tty=/dev/fb0:size=480x800:mmsize=115x170:offset=0 success
[INFO]1970/01/01 00:00:15 set QWS_DISPLAY: transformed:rot0:LinuxFB:mmWidth115:mmHeight170:0 success
```
So it's the embedded fb driver from qt.

**Comment by KaruroChori** [comment]
I am writing a demo in C++ using lvgl, and few other support libraries to do read data from moonraker. It seems to work, but I am having unrelated issues building the project with meson.

As I documented in a different discussion, we have access to the screen via framebuffer `/dev/fb0`, we can disable it via blank `echo 4 > /sys/class/graphics/fb0/blank` (but I am not able to find any control for the backlight), and the touchscreen is `/dev/input/event0`.

 It is likely that we can similarly use the buzzer to make noises if needed, but I have not investigated it further.

**Comment by consp** [comment]
> As I documented in a different discussion

Is this public?

> It is likely that we can similarly use the buzzer to make noises if needed, but I have not investigated it further.
```
# echo 1 > enable
# echo 0 > enable
# pwd
/sys/class/pwm/pwmchip0/pwm6
```
`duty_cycle` and `period` are relevant

note: earplugs might be required.

The play command simply modifies those.

**Comment by KaruroChori** [comment]
No, I don't think it is, but I just reported any piece of information I was able to find out :).

Nice to know we can have the buzzer working as well. Ideally it can be used to signal a finished print, the filament snapping or other hardware issues. Or to output the doom soundtrack.

**Comment by consp** [comment]
> Or to output the doom soundtrack

Why stop at the soundtrack?

**Comment by KaruroChori** [comment]
We also have many motors. We can disable microstepping and interpolation and go polyphonic.

>Why stop at the soundtrack?

:laughing:

**Comment by xblax** [comment]
@consp @khayrulinnikita 
I invited you to the mod repo. I believe most safety issues are already resolved with installing, so probably it should become public soon. If you install the last build, update the printer configs from git before using it.

I will start a discussion there sometime later this weekend if anything else should be resolved before making it public.

Regarding the buzzer: you can look at the 'play' executable from Flashforge. I thing it uses multiple GPIOs for polyphonic tones. In Ghidra I could see that the program already has some tunes built-in ( I think it was Bethoven )

**Comment by consp** [comment]
> I think it was Bethoven

```
tune_error
tune_startup
tune_print_done
tune_makerbot_tv
tune_beethoven_5th
tune_filament_start
tune_pause
tune_sailfish_startup
```

**Comment by xblax** [comment]
> Running X will be a stretch, klipperscreen is designed for hardware an order of magnitude faster (ok, not that much but you get the idea). Anything which uses a bitmapped framebuffer might work. Afaik the application (firmwareExe) is a Qt4.8.6 applet outputting to the fb with the qt embedded display driver, from what I can tell with a quick look.
> 
> The device is a simple 800x960 pixel framebuffer so no technical limitations there. Should be doable (at least if gives nice noise when doing `dd if=/dev/urandom bs=1 count=3072000 > /dev/fb0`)

The actual display is a bit smaller, i think it was 800x480. But you can png to framebuffer image like this
```
convert -size 800x960 xc:none $img -geometry +0+0 -composite -depth 8 bgra:- | xz -c > "./fb/$(basename $img .png).img.xz"
```

**Comment by consp** [comment]
> The actual display is a bit smaller

Yes, see below (480x800), it's double buffer. Doesn't really matter, you don't need the double buffer anyways with the low framerate.

```
[INFO]1970/01/01 00:00:15 set QT_QPA_PLATFORM: linuxfb:tty=/dev/fb0:size=480x800:mmsize=115x170:offset=0 success
```

**Comment by Ross922** [comment]
> > Forgive my ignorance but do you think its possible a company like BTT will come out with a new MB for the 5M Pro if it gains enough traction? That way the weak cpu and MB package can be bypassed etc? Note: I am not familiar with what it takes to put klipper onto this machine nor root and mod for klipper to work on the stock HW, im just asking what is likely an uninformed general question :-)
> 
> I don’t think so, only if in an official collab. The stock CPU is enough to run clipper + moonraker + mainsale + fluidd. There are also hopes that it will be enough to run clipperscreen, then we will get a full-fledged printer

Just imagine this HW on a Rpi3+/4 and a good MB... I got the pro for $439 on Amazon for the ASA capability given the continued issues with K1-series reviews and a recall on my Bambu A1 left me without a printer haha.

**Comment by g992** [comment]
> Just imagine this HW on a Rpi3+/4 and a good MB... I got the pro for $439 on Amazon for the ASA capability given the continued issues with K1-series reviews and a recall on my Bambu A1 left me without a printer haha.

if u have tablet, stock hw is awesome, i use only web interface to control printer

**Comment by jtenniswood** [comment]
I just got my 5M, amazing machine, would love to help test.

**Comment by consp** [comment]
> Running X will be a stretch

Boy was I wrong.

Got X going with GTK3 (and thus GL) and touch. Not working perfectly as expected but the gtk3-demo app runs, albite a bit compressed image wise and some issues with the default touch settings. It needs refinement. KlipperScreen might just work (if we can keep the memory hogging down).

Not even close to public ready so wait a while but there is hope.

Also: no real music, pretty much 8086 PC speaker sounds only and very limited at that. If you spam the pwm driver too much it will just drop dead and ignore you (yes I tried playing the doom music at about 10sample/sec, it stops after about 1s and sounds horrible).

**Comment by KaruroChori** [comment]
Too bad. At least we can use the speaker to make sound effects and use the stepper motors for the soundtrack :D.

**Comment by Sunflex616** [comment]
@xblax don't wanna be THAT guy  but are there any updates regarding an ETA for going public? 
If not I'd really appreciate if I could be added to the repo as well :)

**Comment by consp** [comment]
> Also: no real music, pretty much 8086 PC speaker sounds only and very limited at that. If you spam the pwm driver too much it will just drop dead and ignore you (yes I tried playing the doom music at about 10sample/sec, it stops after about 1s and sounds horrible).

Wrong again, though it's stil very limited and definitely not production ready.


https://github.com/g992/flashforge-ad5m-5mpro-research/assets/6802998/180bec5b-8110-4c68-80f5-921450cad6ca



I'll know it's good enough when there is a copyright strike.

**Comment by Ross922** [comment]
Found this on my recommended: https://www.reddit.com/r/FlashForge/comments/1b1vjgk/klipperscreen_running_on_ad5m/

**Comment by KaruroChori** [comment]
It is the same person before your post.

**Comment by consp** [comment]
> It is the same person before your post.

Internet always goes full circle.

**Comment by Ross922** [comment]
> It is the same person before your post.

Oh haha... their video clip is showing as 0:00/0:00 for me... no video available... my bad :P

**Comment by consp** [comment]
> > It is the same person before your post.
> 
> Oh haha... their video clip is showing as 0:00/0:00 for me... no video available... my bad :P

Might be since it was posted in a private repo first. Has been fixed

**Comment by jtenniswood** [comment]
How's the working going? I'm excited to see klipper running on my new 5M.

**Comment by consp** [comment]
> How's the working going?

No headcrashes yet, so if you consider that good news then it's going great. Currently running a 3h print and monitoring, let's see how it goes. The goal is to get most of the "you can destroy your printer with it" stuff out of the way.

**Comment by DougJoseph** [comment]
> If someone else want's to try this out already, maybe leave a comment here.

Please would you add me as well? I'd love to help test. I have an AD5M and I have root access using the thing from Stone-Time.

**Comment by DomanRulez** [comment]
Could you add me too? I have 5M with few cameras (so far in OctoPrint)

**Comment by Tiwatz** [comment]
Please add me if you want more testers. Got the 5M Pro.

**Comment by nayoshi** [comment]
I would love to participate in testing the firmware, if yall still need more hands on deck :D

**Comment by xblax** [comment]
@DougJoseph @DomanRulez  @Tiwatz @nayoshi 
Thanks for your interest in the mod. We a currently working on the final tasks to make the repository open for public beta. If no more show-stoppers are found it should be ready in a couple of days. Then you are all welcome to all participate in the testing and discussions.

I will announce it here when it's ready.

**Comment by DougJoseph** [comment]
@xblax awesome. Thanks!

**Comment by skylatian** [comment]
Any updates here would be appreciated! I have a 5M Pro and would be more than happy to test :)

**Comment by KaruroChori** [comment]
@consp  & @xblax are completing the documentation for the first public beta release, but all the relevant functionality has been already implemented and tested (up to a certain extent; it is beta for good reasons).
@consp is also working on a pre-alpha branch to fully support klipperscreen. So far it has been reported to work, but there are still some rough edges and the long term stability has not been fully assessed; that feature will not reach a beta status before the initial release.

xblax will just release it once it is ready :).

**Comment by consp** [comment]
> still some rough edges and the long term stability has not been fully assessed; ... beta status before the initial release.

On the mark, but need more testers to get those rough edges out :) Won't be long, patience is a virtue! (and makes us not screw up :roll_eyes: )
Already for everyone to prepare themselves: Make sure your slicer is set up correctly, the default start/stop orcaslicer settings from FlashForge will NOT work.

**Comment by mBlinkii** [comment]
Hello, I am looking forward to your work. How well do Mainsail and KlipperScreen run on the original board? I'm thinking of replacing the whole board and using an RPI Klipper and Mainsail.

**Comment by glizzyglizz69** [comment]
I would like to become a tester - Will the documentation provide instructions on how to install the firmware on the printer's mainboard? Also, when said documentation is released, will we be updated here in this forum? I'm subscribed and receive updates here, just want to make sure I'm able to get everything. Thanks!

**Comment by xblax** [comment]
The waiting has come to an end, finally :tada:

I have set the repository https://github.com/xblax/flashforge_adm5_klipper_mod to public and everybody who's interested can now have a look and help testing. *Please note that it's still beta.*

That also means that we no longer have to hold this issue hostage for public discussions of the mod. I will close it and invite everyone continue the discussion in the [Adventurer 5M Klipper Mod v00.03-beta Release Announcement](https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/49) or the other open topics in the mod repository.

**Comment by TMTYD** [comment]
hi

Does that also work with the AD5X?

**Comment by TMTYD** [comment]
hi

Does that also work with the AD5X?

**Comment by g992** [comment]
> hi
> 
> Does that also work with the AD5X?

As I can see from the video (https://youtu.be/5h_mIf8kq_M?t=56), there's a different processor, so the mod won't work, requires adaptation and rebuild of the OS (Buildroot system)

**Comment by cfelicio** [comment]
https://github.com/Tina-Linux/linux-5.4 is gone, anyone have another source?
